### PR TITLE
[GlobalIsel][NFC] Refactor ext of trunc

### PR DESCRIPTION
**Pull request by tschuett**

### CODE REVIEW

**Review by arsenm — APPROVED** [review]

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 1]
```diff
@@ -735,18 +735,20 @@ def redundant_sext_inreg: GICombineRule <
 // the destination type.
 def anyext_trunc_fold: GICombineRule <
   (defs root:$root, register_matchinfo:$matchinfo),
-  (match (wip_match_opcode G_ANYEXT):$root,
-         [{ return Helper.matchCombineAnyExtTrunc(*${root}, ${matchinfo}); }]),
-  (apply [{ Helper.replaceSingleDefInstWithReg(*${root}, ${matchinfo}); }])
+  (match (G_TRUNC $input, $x),
```
Comment (jayfoad) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:738]: Does this still match if there is a G_ASSERT_ZEXT in between? If so, how? If not, why doesn't it regress the tests added in #74125?

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:738]: No. The output `input` of the G_TRUNC is the input `input` of the G_ZEXT.

Comment (jayfoad) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:738]: `test_combine_anyext_trunc_with_hint` in `test/CodeGen/AArch64/GlobalISel/combine-ext.mir` tests this sequence:
```
    %1:_(s32) = G_TRUNC %0(s64)
    %2:_(s32) = G_ASSERT_ZEXT %1(s32), 11
    %3:_(s64) = G_ANYEXT %2(s32)
```
I would expect your new matcher to not match this. How does it work?

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:738]: The Linux VM ran through without issues.

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:738]: It makes an optimization for COPY, but it did never support G_ASSERT_ZEXT.
https://github.com/llvm/llvm-project/blob/30d0850e0f780b17a37522e6503c98ebe197c5fa/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L2438

Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:738]: No tests failed, but it is not NFC. The copy optimization does not hit anymore, which includes G_ASSERT_ZEXT.

Comment (jayfoad) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:738]: @Pierre-vh am I right in thinking that a pattern like this:
```
  (match (G_TRUNC $input, $x),
         (G_ANYEXT $root, $input):$mi,
```
use GIM_RecordInsn**IgnoreCopies** to check that `$input` matches? (I can't see exactly where this is implemented, nor find any discussion of it.)

If so then I think this patch is probably fine and you can remove most of the C++ matching code. All you need to check is that the source type of the trunc matches the dest type of the ext.

**Pull request closed without merging**